NXxas refactoring

woutdenolf opened the pull request:
WIP: everything will be moved to contributed_definitions once this is up for review.

**Contributing**: PR to the main branch of https://github.com/XraySpectroscopy/nexus_definitions (merge commits will be removed regularly).

**Rendering**: https://hdf5.gitlab-pages.esrf.fr/nexus/nxxas/classes/applications/NXxas_new.html

**Example files**: https://pynxxas.readthedocs.io/

**NXDL questions**:

    - how to re-use an enum
    - when enum field X has a specific value, make field Y required OR how to define enum properties
    - possibly link to
    - how to make a field required in a base class

woutdenolf committed:
first refactored NXxas template

woutdenolf committed:
improve NXelectron_level

woutdenolf committed:
use X-ray level

woutdenolf committed:
add more documentation

lukaspie commented:
Hi @woutdenolf, I came across this PR by chance and noticed that you are working on a base class for describing an electron level. In the context of photoemission spectroscopy, we in FAIRmat have also been working on something very similar: https://github.com/FAIRmat-NFDI/nexus_definitions/blob/fairmat/contributed_definitions/NXelectron_level.nxdl.xml. In fact, some of the base class we developed was originally adapated from another XAS-related PR: https://github.com/nexusformat/definitions/pull/1293. Would you be open to harmonizing the concept that we were developing with your approach here? I think the two approaches are already very similar right now.

woutdenolf commented:
@lukaspie Yes since we are doing the same thing we should try to converge. We are currently discussing this very topic: https://github.com/XraySpectroscopy/nexus_definitions/pull/2. Feel free to get involved.

The XAS working group will work in https://github.com/XraySpectroscopy/nexus_definitions. It would be helpful to add comments in that repo. This PR will be a draft and is subject to lots of changes until we are finished.

woutdenolf commented:
> another XAS-related PR: https://github.com/nexusformat/definitions/pull/1293

This was the first attempt before we created the XAS working group. It will most likely be closed.

whs92 commented:
@woutdenolf after asking a question to the nexus mailing list and reading:

https://github.com/nexusformat/definitions/issues/1011
https://github.com/nexusformat/definitions/pull/1293

I arrived here :)

I have had a look at the proposal you made to the [NXxas_new](https://nxxas-hdf5-nexus-7ba1891a1aaa2d29580131298dc337bab2b3e5de32b6ef.gitlab-pages.esrf.fr/classes/applications/NXxas_new.html) 

I am particularly interested in the topic raised in #1011 of adding the ability to work with multiple detectors. I see that you've added the entry `intensity` that is still tied to a single mode (TFY, PFY, TEY etc). How might I add other signals acquired at the same time? How might I add more than one PFY signal?

Do you plan to include the suggestions from @padraic-shafer in https://github.com/padraic-shafer/definitions/blob/1011-multixas/contributed_definitions/NXmultixas.nxdl.xml

When is the next planned meeting of the working group? Can I come along?

woutdenolf commented:
Currently we only support a single XAS spectrum. I'm very much open to support multi-detector and/or multi-scan. You are very welcome to join the XAS working group. I'll contact you directly.

woutdenolf commented:
> I'll contact you directly.

Well you'll have to contact me because you have no contact info.

mretegan committed:
Remove NXxray_level base class

mretegan committed:
Use enumerations

benajamin commented on applications/NXxas_new.nxdl.xml:
```diff
@@ -0,0 +1,230 @@
+                <field name="name"/>
+                <field name="probe">
+                    <enumeration>
+                        <item value="x-ray"/>
```
If you want to allow the "electron energy loss" mode, then you should also allow an "electron" probe type.

benajamin commented on applications/NXxas_new.nxdl.xml:
```diff
@@ -0,0 +1,230 @@
+            </enumeration>
+        </field>
+        <group type="NXelement" name="element">
+            <doc>Excited element</doc>
```
What if I want to do a scan that includes multiple absorption edges (e.g. Carbon K-edge and K L-edge)? Can I specify multiple elements?

benajamin commented on applications/NXxas_new.nxdl.xml:
```diff
@@ -0,0 +1,230 @@
+                </item>
+                <item value="electron yield">
+                    <doc>
+                        TODO total or partial electron yield
```
Does this include Auger electron yield?

benajamin commented on applications/NXxas_new.nxdl.xml:
```diff
@@ -0,0 +1,230 @@
+                      </dimensions>
+                </field>
+            </group>
+            <group type="NXdetector" name="i0" optional="true">
```
The i0 should have a corresponding NXmonitor group. There are also many ways of measuring an i0 that is useful to document in the file.

ounsy commented:
> @woutdenolf after asking a question to the nexus mailing list and reading:
> 
> #1011 #1293
> 
> I arrived here :)
> 
> I have had a look at the proposal you made to the [NXxas_new](https://nxxas-hdf5-nexus-7ba1891a1aaa2d29580131298dc337bab2b3e5de32b6ef.gitlab-pages.esrf.fr/classes/applications/NXxas_new.html)
> 
> I am particularly interested in the topic raised in #1011 of adding the ability to work with multiple detectors. I see that you've added the entry `intensity` that is still tied to a single mode (TFY, PFY, TEY etc). How might I add other signals acquired at the same time? How might I add more than one PFY signal?
> 
> Do you plan to include the suggestions from @padraic-shafer in https://github.com/padraic-shafer/definitions/blob/1011-multixas/contributed_definitions/NXmultixas.nxdl.xml
> 
> When is the next planned meeting of the working group? Can I come along?

@woutdenolf 
I also got this request from one of beamline scientists at SOLEIL

ounsy commented:
https://github.com/nexusformat/definitions/pull/1352#issuecomment-1994588748
@whs92 
@woutdenolf 
I also got a request for that from one of our beamlines at SOLEIL

mretegan committed:
Rename absorption_edge to edge

mretegan committed:
Remove emission_lines

mretegan committed:
Modifications after the meeting

mretegan committed:
Remove unused fields

Co-authored-by: Wout De Nolf <wout.de_nolf@esrf.eu>

mretegan committed:
Merge pull request #4 from XraySpectroscopy/use-enumerations

Use enumerations for the absorption edge and emission lines

mretegan committed:
Add initial text for XAS absorption modes

mretegan committed:
Added more detection modes

mretegan committed:
List the modes in the documentation

mretegan committed:
Remove empty line

newville committed:
Merge pull request #5 from XraySpectroscopy/update-xas-modes

Add documentation and extend the XAS modes

PeterC-DLS reviewed (commented):
With `NXedge` and `NXemission_lines`, is it better to have an atomic orbital (in X-ray notation rather than spectroscopic notation) and transition with the latter comprising an initial and final orbital?

DanPorter commented:
Hi all, I was asked by @PeterC-DLS to have a look at this as a few beamlines I'm working with at Diamond are using sub-entries with the old NXxas specification. I have a couple of comments:

1. There is a comment from @benajamin about multiple edges - this is a valid point. I think there should be an additional base class NXabsorption_edge that can contain NXelement and NXedge, the name could be an enum like "Co L3". The NXentry could then contain as many NXabsorption_edge fields as required. This way multiple edges/ elements could be specified, the principle or intended edge would be the first used. In my mind this would look something like:
```
/entry/definition = "NXxas_new"
/entry/mode:NXmode
/entry/Co_L3:NXabsorption_edge
/entry/Co_L3:name = "Co L3"
/entry/Co_L3/element: NXelement = "Co"
/entry/Co_L3/edge_type: NXedge = "L3"
/entry/Co_L3/orbital = "2p3/2"
/entry/Co_L3/energy = 778.1 @units="eV"
```
As @PeterC-DLS mentioned, this could also be expanded to allow different notations and the specification of specific orbital transitions.
I've written an example enumeration script for NXedge with name type "element edge" [here](https://gist.github.com/DanPorter/13791d09609593a256d8ce86cdbfda6c), in case this is useful.

2. The NXdata group should probably contain optional links to the monitor and to fields in NXprocess, for example (I/I0) or log(I/I0). It should be more clear in the application definition that calculated experimental spectra like this should be stored in NXprocess (assuming this is correct).
3. The "calculated" field should probably be renamed "simulated" or, inversely, "measured". 'calculated' could mean calculating a ratio of measurement signals, and I assume this field is to determine if the spectra has originated from simulation software. In which case, should there be some optional fields, analagous to NXinstrument, that capture what software and what inputs were used?

newville commented:
@DanPorter 
Point 1: It would be helpful for NeXus to have standard classes for Element and either X-ray Level or Electron Level.  We in the XAS community spent a lot of time trying to come up with such a proposal and gave up and decided it was not our fight.  See @PeterC-DLS' comment for example.  We spent a lot of time lost in such discussions.  We concluded that "Element" is a string of an atomic symbol, and "Edge" is a string of the principal edge measured.   Yes, measurements can be "L3, L2", or "M5, M4, M3" -- those would be allowable.   

I would not be in favor of an "element_edge" like "Co L3".   We will want to be able to search on Element and also on Edge, and should not expect all downstream software to parse such strings.   

Point 2: Yes, the intention is that the `process` entry should describe how the "rawdata" data gets transformed into the `intensity` array (and perhaps `energy`, if that is not directly measured).  This should describe what math is done on the raw data arrays (or perhaps other NXxas entries).  It will likely be more complex than  `-log(Column3/Column2)`, as it might apply deadtime corrections and sum multiple channels from the "rawdata" arrays.  
And, since the `intensity` to be presented (and plotted) is the "lightly processed normalized mu(E) data", the processing required to get to `intensity` should include some background subtraction, typically the slope/intercept of a linear background in energy, and the normalizing edge_jump used.

Point 3: "Calculated" could be the sum or mean of multiple other spectra, or the result of a fit or some other analysis.  But, if the ratio of different signals in the "raw data" within a single measurement, then every spectrum would be a calculation - no one measures one "signal" that is "normalized mu(E)" or even "unnormalized mu(E)".     But also, yes, including the details of the calculation or analysis done in the process section would be helpful and/or expected.

DanPorter commented:
Thanks for your responses @newville. I certainly don't want to re-open old arguments!

I just thought of another comment whilst whilst working on some XMCD spectra:

4. When looking for magnetic signals we subtract different spectra based on polarisation, so we need to know the incident beam polarisation. Currently we store a string defining 'lh', 'lv', 'pc', 'nc' polarisation states in `/entry/instrument/id/polarisation` and also put the stokes vector in /`entry/sample/beam/incident_polarization`. Can at least the NXsample>NXbeam>incident_polarisation field be included as an optional field?

woutdenolf committed:
Merge branch 'nexusformat:main' into main

mkuehbach committed:
Revisiting non-FAIRmat constructive solid geometry in light of FAIRmat proposal  (#1532)

* Added references and relaxed constraint that only NXquadric or NXoff_geometry can be used to define geometries as instances

* Renamed parameters to parameter as typically field names should stick to singular

* Promoted NXquadric to proper base class

* Promoted dependent base class to a proper base class

* Relaxed constraints to allow also other geometries as instance data

* Promoted NXcsg to a proper base class

* Linting

* Trying to figure out how to add urls properly

* Fixed test

* Removed the old NXsolid_geometry as it this issue proposes to promote that class to a base class

* Addressed @PeterC-DLS comments

* Reset parameters how it was

* Fixed typo

* Implementing feedback from 2025/07/07 telco

* Update copyright year to 2025

---------

Co-authored-by: mkuehbach <markus.kuehbach@physik.hu-berlin.de>

woutdenolf committed:
removed unused namespace prefixes xs and ns (#1372)

woutdenolf committed:
NXstress contribution: stress-strain analysis (#1039)

* contribution for stress-strain analysis

* ILL contributions

* Updated NXstress

* peaks are NXreflections instead of NXdata

* units cannot be defined by symbols

* beam_intensity_profile becomes NXbeam instead of NXcollection

* use NXnote groups with a required calibration_type field for calibration

* add sin2psi diffraction type and qmagnitude in case cUnits is not enough (like sin2pi)

* qmagnitude is redundant (it is center with type momentum transfer) and can cause confusion

* fix NXsample transformations

* fix indentation

* fix link name

* use enumerations

* rename diffraction_type to processing_type, use enumeration and add more descriptions

* fix PDF build

* fix math blocks

* Update NXstress.nxdl.xml

* Update contributed_definitions/NXstress.nxdl.xml

Deleted the energy dispersive information.

* Update contributed_definitions/NXstress.nxdl.xml

Change back again the angstrom symbol

* Update contributed_definitions/NXstress.nxdl.xml

Co-authored-by: Arnold Paecklar <61598106+aapaecklar@users.noreply.github.com>

* fix misconception about energy/wavelength

* Update contributed_definitions/NXstress.nxdl.xml

Co-authored-by: Aaron S. Brewster <asbrewster@lbl.gov>

* fix missing peak_parameters groups

* use NXparameters where appropriate

* cannot be distinguished by a reader when the names are optional

* fix typo

* move NXstress to applications

* move NXstress to applications

* Replace all optional with minOccurs and maxOccurs and use nameType="any" where appropriate

* copy symbols inherited from NXsample

* add missing maxOccurs

* update year

* snake case for all symbols

* Update applications/NXstress.nxdl.xml

Co-authored-by: Peter Chang <peter.chang@diamond.ac.uk>

* Apply suggestions from code review

Co-authored-by: Peter Chang <peter.chang@diamond.ac.uk>

* replace phrase references with links

* fixup NXdetector distance

* depends_on directly under NXsample

* delete blank line

* fig_log group is now FIT/description

* gauge volume coordinate system

---------

Co-authored-by: Arnold Paecklar <61598106+aapaecklar@users.noreply.github.com>
Co-authored-by: Aaron S. Brewster <asbrewster@lbl.gov>
Co-authored-by: Peter Chang <peter.chang@diamond.ac.uk>

newville commented:
@woutdenolf Thanks for updating this. Any guidance on the delay on merging this?  I'm sort of at a loss here.

DanPorter commented:
Hi all, at a previous meeting I promised some example data and I'd forgotten to share it with you! Here is a repo with some example scans doing XAS and XMCD, where I've tried to implement the new format:

https://github.com/DanPorter/xmcd_analysis_example

You can click on the links in the readme to view the files in [myHDF5](https://myhdf5.hdfgroup.org/).

The files also contain processed components described with NXprocess. I found that there are lots of different ways of interpreting how this should be implemented, but the way I've done it provides useful default plotting and what I felt is a full description of the steps taken, with different steps stored as different NXdata groups and different elements in the NXprocess group.

mretegan commented:
Hi all,

@DanPorter thanks a lot for providing the formatted files. I don't think we discussed using this definition to describe XMCD, or any other type of dichroic measurements for that matter. I would be more in favor of not adding too much initially so that we have something finished for XAS; this has already taken more time than expected. This week I can update the definition with the information from the [modes](https://docs.google.com/document/d/1M-9ViaEiaR6fUwb5IffQILodOSTvJq5s/edit)

Regarding the “calculated” field, it is meant to indicate whether the data is from a theoretical calculation, so its description could be improved.

samtygier-stfc committed:
Typo: NXtomoproc nX -> nY (#1243)

samtygier-stfc committed:
Generalise datatype of NXtomoproc data to NX_NUMBER (#1244)

Tomography software typically outputs floating point arrays for reconstructed data, so it is useful to store without further transformation.

Closes #1241

As discussed in Telco 20230213

zdemat committed:
NXinsertion_device: wavelength_shifter  (#1587)

* `wavelength_shifter` added to the list of `type` of `NXinsertion_device`

* Create pages.yml

* adding docs string for insertation device type

* fix docs entry position

* Update base_classes/NXinsertion_device.nxdl.xml

Co-authored-by: Peter Chang <peter.chang@diamond.ac.uk>

* Delete .github/workflows/pages.yml

Reverting dae1fd6. .github/workflows/pages.yml should ot be included in this PR.

---------

Co-authored-by: Sonal R. Patel <sonalpatel.982@gmail.com>
Co-authored-by: Peter Chang <peter.chang@diamond.ac.uk>

lukaspie committed:
add spell checker in CI/CD and Makefile (#1582)

Co-authored-by: mkuehbach <markus.kuehbach@physik.hu-berlin.de>

lukaspie committed:
Fairmat 2025: changes to contributed definitions (#1427)

* modify moving of existing docs

* remove commented CI code that is not needed anymore

* Opened NXem_conventions for usage in methods other than electron microscopy

* Consolidated based classes related to NXem_conventions into NXcoordinate_system_set because that is what these conventions only talk about

* Refactored NXmicrostructure_score_config

* Refactored NXmicrostructure_gragles_config

* Even make html runs through locally

* remove _source build artifacts before docs deployment

* Finished NXmicrostructure_score_results and hope that commenting out the rollett-holm formula fixes the build error with latex, locally at least it does

* Fixed the too stiff name of microstructure1

* Initial version of NXmicrostructure_gragles_results

* Fixed grammar issue

* Finished NXmicrostructure_gragles_results

* Implemented suggestions from @lukaspie

* nxdl

* NXdata linking of signal (#205)

* Add depends for data field

* Use `@reference`

* use reference attribute in NXmpes and NXmpes_arpes

* change text style

* fix duplicated in reference attribute

---------

Co-authored-by: Lukas Pielsticker <50139597+lukaspie@users.noreply.github.com>

* Roll back changes to NXmpes/instrument/pressure_gauge

* fix 2theta and use NXobject instead of collection.

* reproducibity check.

* add radius to NXenergydispersion

* Xrd modification (#274)

* Including scattering vector components perpendicular and parallel to the surface of the sample those are mainly used for RSM diffraction technique.

---------

Co-authored-by: Florian Dobener <florian.dobener@physik.hu-berlin.de>

* Implementing suggestions from lukaspie

* Refactored NXcoordinate_system_em_ebsd into NXem_ebsd

* Fix naming bug for monochromator_em

* Bug fixes in images and spectrum and renaming of groups, refining precision of concept image and how this is connected to tilings

* Add circuit to component

* Adding routine tests for py3.12

* Added circuit and corrected laziness error

* Editing NXem for aberrations in event data

* Extension of Doc for Optical spectroscopy and Raman

* Update webpage for optical spectroscopy

* Further simplification of where and how to store aberrations

* change ellips name to optical spec

* Use something from lensID

* Correcting units for complex aberrations

* Removed mandatory model for aberrations as clear for naming conventions used

* Reference Rafal for aberration function

* Fixed time(NX_TIME) bug

* fix_renaming_of_ellipsometry_to_optical_spec

* add missed links

* add missing ####  in title...

* Clarified issues with depends_on, fixed typo pID for pfID in NXem_ebsd

* Hooked in examples for usage of NXem_ebsd and NXem_eds into NXem

* Last cleaning for depends_on on this feature branch

* Initial draft of NXmpes_xps sub app-def

* Prepare NXmpes_xps for extending NXmpes

* Removed information already given in base classes

* Removed some unneeded comments

* Add description of instrument orientation

- Angle between beam and sample normal-> needed for determination of probing depth
- Angle between beam and analyzer axes -> needed for data analysis

* Move orientation of instrument parts

Instead of having two fields on the instrument for the angle between beam and sample and the angle between beam and analyser, this information is stored in depends_on and NXtransformations in NXinstrument and NXsample, respectively.

* Update transmission function information

 - The transmission function contains an additional NXnote field "file" describing the file in which the TF data is stored.

* Update documentation, add comments for discussion

* Initial draft of NXmpes_xps sub app-def

* Prepare NXmpes_xps for extending NXmpes

* Removed information already given in base classes

* Removed some unneeded comments

* Add description of instrument orientation

- Angle between beam and sample normal-> needed for determination of probing depth
- Angle between beam and analyzer axes -> needed for data analysis

* Move orientation of instrument parts

Instead of having two fields on the instrument for the angle between beam and sample and the angle between beam and analyser, this information is stored in depends_on and NXtransformations in NXinstrument and NXsample, respectively.

* Update transmission function information

 - The transmission function contains an additional NXnote field "file" describing the file in which the TF data is stored.

* Update documentation, add comments for discussion

* fix NXmpes yaml

* update to current NXmpes status

* refine NXtransformations

* add coordinate system and coordinate_system_set

* Adds mpes_xps to mpes structure

* rename to NXxps

* initial attempt at modeling XPS peak fitting

* add lineshape to NXpeak

* update lineshape in NXpeak

* add NXbackground to NXfit_region

* add fitting method to NXpeak_model

* remove NXtransformations from NXcoordinate_system

* restructure NXfit with background and peaks

* change to NXcoordinate_system_set, use NXfit in NXxps

* add fit_function and parameters

* use NXfit_function for peaks and backgrounds

* use NXdata in NXfit

* use new peak fitting in NXxps

* add missing class name in NXxps

* remove underscores from variable names in NXfit

* update nyamls

* switch axes docs in NXfit

* use old nyaml dim notation for now

* fix math notation

* make nxdls

* regenerate with nyaml==0.0.8

* make NXfit, NXpeak multi-dimensional

* redefine XPS coordinate system, small changes to XPS peak fitting

* required energy axis in XPS

* fix dimensions issues

* regenerate NXxps

* make energy an NX_NUMBER

* remove unneeded depends_on enumerations

* align symbols across all fit-related definitions

* remove single and repeated parameters in fit

* fix dimensions in NXpeak and NXfit_background

* use globlal_fit and error_function in NXfit, fix dimensions

* docstring fixes in NXfit

* NXfit extends NXprocess

* fix math notation in NXfit_function

* fix link in NXfit_function

* change requiredness for formulas in XPS fit

* small fixes for links

* fix issues with manual build

* fix math notation

* more math fixes

* regenerate NXxps and NXpeak nxdl files

* fix rebasing issues

* make NXpeak spectrum-independent

* fix rebase conflicts

* fix nyaml-nxdl inconsistency

* add notes from both workshops

* added make nxdl and make local

* lowercase enumeration, source_type todo added, doc for ellipsometry_type

* HOW LONG? does it take to learn to make nxdl and local before commiting >.<

* Fixes in EM

* Merging NXem_adf into NXem_img, enabling microstructure reconstructions as dependants of specific method results, introduction of NXatom_set to discuss generalization of NXion towards enabling descriptions of groups of atoms, also to lias with FAIRmat Area C

* Removal of non-standardized assumption that i, j, k can be used as counter variables in case one does not know the actual dimensions but one is sure it is not a scalar

* add missing elements in ellipsometry from specialized NXlens_opt

* Add todo notes for beamsize and shape  description

* typo and note for possible rework/reconsideration of NXfabrication

* Suggestions from lukaspie

* Fixes apm

* Added undefined enum in NXidentifier

* use NXidentifier in NXsample and NXfabrication

* rename sample name in NXoptical spectroscopy

* make ellipsometer_type optional

* NXidentifier in NXoptical_spectroscopy instrument

* remove NXsample name docs from NXoptical_spectroscopy

* add serial_number to NXfabrication

* use NXidentifier in NXmpes

* use NXidentifier consistently across all our contributed definitions

* decode binary strings in nxdl_utils.py

* add function for decoding

* use NXidentifier in NXentry

* small changes to NXbeam and NXmonochromator

* typo fixes

* proper string decoding

* remove unneeded imports

* black formatting

* remove code and move to pynxtools

* remove unneeded requirements

* escape black check

* ignore flake8 issue

* isort imports

* use less strict version of str decoding

* slightly enhanced version of str decoding, test

* remove flake error catching

* reset requirements.txt

* reformatting

* remove strict type checking

* add support for list of bytes

* formatting

* add pulse_delay to NXbeam

* recreate nyaml file

* regenerate nyaml files

* various small fixes are merging main and fairmat

* fix string No in NXelectron_level

* remove unused MANIFEST.in

* manually reset NXem yaml file

* format base classes and applications in accordance to existing NIAC version

* revert changes to NXtransformations docs

* add CSS tweak to hide summary

* add css file

* Suggestions from Lukas

* fix dev_tools tests
black code style

* remove use of details_summary_hide

* change refs in NXarpes

* move doc bundles to applications and base_classes

* Reverting the proposal that NXmicrostructure is not just proposed for contributed but implicitly as a dependency of NXem, NXem_ebsd, NXem_img proposed for voting and acceptance

* Revert one more microstructure

* move new definitions to application and base_classes

* small update to mpes-structure

* update category for NXxrd.nxdl. (#296)

* remove fairmat-specific files

* remove nyaml files

* move NXdata_mpes* classes back to contributed

* revert small changes to the dev_tools

black formatting

isort dev_tools

* change docs in NXcircuit

* pull out modifications for fairmat-2024-contributed

* bring in multi line doc handling in dev_tools

* remove renamed classes

* bring in latest NXxrd from fairmat repo

* bring in changes from fairmat repo

* bring in changes to optical spectroscopy classes

* modify dispersion models

* bring in NXmicrostructure from fairmat repo

* small changes to NXtransmission

* bring in NXspatial_filter from fairmat repo

* bring in NXmatch_filter from fairmat repo

* bring in NXisocontour from fairmat

* bring in NXdelocalization from fairmat

* bring in NXem_calorimetry from fairmat

* remove NXlab* classes

* bring in NXsubsampling_filter from fairmat branch

* bring in NXsimilarity_grouping from fairmat branch

* bring in NXapm_paraprobe classes

* bring in NXapm_compositonspace* classes from fairmat

* remove NXamplifier (part of SPM proposal)

* fix various issues with the documentation build

* typo fixes

* Clarify situation wrt to NXclustering

* Fix error in cgms docs

* Removed NXcs_ classes that in the process of the acceptance of NXapm have were refactored as base classes, e.g. NXcs_computer

* Capitalization, typos and header docstring for electrostatic kicker

* Capitalization, typos and header docstring for magnetic kicker

* Capitalization, typos and header docstring for quadrupole magnet

* Capitalization, typos and header docstring for separator

* Capitalization, typos and header docstring for spin_rotator, substance, and xpcs

* review comments

* Latest development stage of FAIRmat for the contributed definitions for atom probe and microstructure

* Reset NXtransmission and NXoptical_polarizer to commits before today

* Sync up that the hash rule for nuclid_hash defined in NXdelocalization and NXatom match up

* Fix typos in dispersive material

---------

Co-authored-by: mkuehbach <markus.kuehbach@physik.hu-berlin.de>
Co-authored-by: mkuehbach <markus.kuehbach@hu-berlin.de>
Co-authored-by: Florian Dobener <florian.dobener@physik.hu-berlin.de>
Co-authored-by: Rubel <rubel.mozumder@outlook.com>
Co-authored-by: RubelMozumder <32923026+RubelMozumder@users.noreply.github.com>
Co-authored-by: Markus Kühbach <mkuehbach@users.noreply.github.com>
Co-authored-by: Ron Hildebrandt <RonHildebrandt@uni-leipzig.de>
Co-authored-by: Ron <139139971+RonHildebrandt@users.noreply.github.com>
Co-authored-by: rettigl <l.rettig@web.de>

lukaspie committed:
Fairmat 2024: groupings in manual (#1428)

Co-authored-by: mkuehbach <markus.kuehbach@physik.hu-berlin.de>
Co-authored-by: Markus Kühbach <mkuehbach@users.noreply.github.com>

lukaspie committed:
Resolve "Invalid type NX_DIMENSIONLESS in NXstress" (#1597)

Co-authored-by: Lukas Pielsticker <you@example.com>

woutdenolf committed:
NXazint contribution (#1577)

* initialize the NXazint contribution

* parse MAXIV proposal

* Renaming momentum transfer magnitude to scattering vector magnitude

* separating out the 2D case and minor refactoring

* 2d NXazint example ala Danmax

* Update NXazint2d.nxdl.xml

Updated based on implementation at DanMAX. Changed to US spelling. Lower case spelling for long names.
Long name for 'norn' must be changed!

* Update NXazint2d.nxdl.xml

Added I_errors field.

* Add nEtaEdge in symbols

Fixed symbols by adding nEtaEdge

* Align 1D and 2D definition

Through update of the NXazint1d definition based on the NXazint2d definition

* Remove nImg from norm

Removing the nImg value from norm and decrease the rank

* long name for norm

added a descriptive long name for the norm

* added long name for norm

added a descriptive long name for the norm

* Fixed syntax error in line 46

* FIX syntax error in line 155

* Update NXazint2d.nxdl.xml

Change units for azimuthal_axis to NX_ANGLE

* capitals, monitor, fix mention of eta bins

* stupid mistake in monitor dimensions

* monitor data

* Update NXazint1d.nxdl.xml

Changed NX_CHAR types to NX_ANY

* Update NXazint2d.nxdl.xml

Changed the unit types deom NX_CHAR to NX_ANY

* update 2d align with 1d

bold for some field names, add normalisation boolean

* fix typos in 2d

* add links in description

* NXazint1d without explicit NXsubgroup and polished

* indent cleaning

* 1d: \t replaced by 2 spaces

* 1d: commas in NXentry description

* NXazint2d with NXsubentry removed

* NXazint2d: indent and white chars

* adding NX_WAVENUMBER alternative for NX_PER_LENGTH

* Fix mostly typos in 1D case.

Co-authored-by: Peter Chang <peter.chang@diamond.ac.uk>

* Fix typos also in 2D case

* doc for normalization_applied and adding optional monitor_applied

* adding fixed/restricted units in monochromator

* moving NXazint(s) to applications definitions

* cosmetic changes in NXazzint(s) documentation

* NXazint(s) - only changes in the documentation of the corrections: 1) removed comments using undefined notation, some statements made more definitive, adding a Note label

* Making the note about normalization strategy more explicit.

* making 'reduction' field lowercase

* Update applications/NXazint1d.nxdl.xml

Co-authored-by: Peter Chang <peter.chang@diamond.ac.uk>

* Update applications/NXazint1d.nxdl.xml

Co-authored-by: Peter Chang <peter.chang@diamond.ac.uk>

* Update applications/NXazint1d.nxdl.xml

Co-authored-by: Peter Chang <peter.chang@diamond.ac.uk>

* Remuving capitalized names of NXclass instances where not needed.

* adding word 'effective' in the long description of norm

* restructured documentation of normalization_applied

---------

Co-authored-by: Zdenek <zdenek.matej@maxiv.lu.se>
Co-authored-by: Paul Bell <paul.bell@maxiv.lu.se>
Co-authored-by: mrvj <mads@chem.au.dk>
Co-authored-by: Frederik <47137695+fgjorup@users.noreply.github.com>
Co-authored-by: Zdeněk Matěj <zmatej@centrum.cz>
Co-authored-by: Zdenek <32390711+zdemat@users.noreply.github.com>
Co-authored-by: Peter Chang <peter.chang@diamond.ac.uk>

PeterC-DLS committed:
Update NXDL version for release

PeterC-DLS committed:
Upload PDF of v2025.11

Also, update and fix release note creator script, tweak base class text
to not use too long footnote (otherwise PDF output has ugly overlap)

PeterC-DLS committed:
Fix CI to allow deployment and uploads

PeterC-DLS committed:
Update CHANGES with release v2025.11

PeterC-DLS committed:
Recategorise some application definitions for #1602

Move tomography definitions to a new imaging category and move scattering definitions to diffraction

woutdenolf committed:
datetime.timezone.utc works for all modern python versions (3.2 -> 3.13+). (#1593)

PeterC-DLS committed:
Expose deprecated attributes in docs for #1605

mretegan committed:
Merge branch 'nexusformat:main' into main

mretegan committed:
Merge remote-tracking branch 'upstream/main'

mretegan committed:
Update the list of emission lines (#8)

mretegan committed:
Update edges (#9)

mretegan committed:
Merge branch 'nexusformat:main' into main

mretegan committed:
Add .readthedocs.yaml

mretegan commented:
The two options for the updated application definition described above and in [Telco 20260211](https://www.nexusformat.org/Telco_20260211.html) are implemented below.

**30/04/2026 The first proposal was retained.**

1. Inheritance - `NXxas` is a parent class. Each mode is a separate application definition that extends `NXxas` and adds mode-specific fields.

```
NXobject
 `-- NXxas (parent)
      |-- NXxas_trans
      |-- NXxas_tfy
      |-- NXxas_pfy
      |-- NXxas_herfd
      |-- NXxas_tey
      |-- NXxas_pey
      `-- ...
```
Rendered documentation: https://nexus-definitions.readthedocs.io/en/xas-using-inheritance/classes/contributed_definitions/NXxas.html

2. Composition - a single `NXxas` application definition. Each measurement mode is a base class that extends `NXobject` (e.g., `NXherfd`) included as a group. The mode group carries all mode-specific metadata and optional instrument details.

```
NXobject
  |-- NXtrans  
  |-- NXtfy  
  |-- NXpfy  
  |-- NXherfd 
  |-- NXtey 
  |-- NXpey  
  `-- ...

NXxas (single application definition, extends NXobject)
  `-- uses any of the above modes
```
Rendered documentation: https://nexus-definitions.readthedocs.io/en/xas-using-modes/classes/applications/NXxas.html

mretegan committed:
Merge remote-tracking branch 'upstream/main'
